kytos/core/controller.py

```diff
@@ -1022,3 +1029,43 @@ def _try_to_fmt_traceback_msg(self, message: str, counter: Counter) -> str:
                 f"{message}\nFull KytosEventBuffers counters: {counter}"
             )
         return message
+
+    def get_link_or_create(self, endpoint_a: Interface, endpoint_b: Interface) -> tuple[Link, bool]: # SAFE FUNCTION
+        """Get an existing link or create a new one.
+
+        Returns:
+            Tuple(Link, bool): Link and a boolean whether it has been created.
+        """
+        with self.links_lock:
+            new_link = Link(endpoint_a, endpoint_b)
+
+            if new_link.id in self.links:
+                return (self.links[new_link.id], False)
+
+            self.links[new_link.id] = new_link
+
+        return (new_link, True)
+    
+    def get_link(self, link_id: str) -> Optional[Link]:
+        """Return a link by its ID.
+        
+        Returns:
+            Optional[Link]: Link if found, None otherwise.
+        """
+        return self.links.get(link_id)
+
+    def get_links_from_interfaces(
+        self,
+        interfaces: Iterable[Interface]
+    ) -> list[Link]:
+        """Get a list of links that matched to all/any given interfaces."""
+        links_found = []
+        with self.links_lock:
```

Comment:
We either use `self.links_lock` or a shallow copy of the links, here you have both. Le's keep as the original code with only the shallow copies for now. 

Also, I just realized that we are also making unnecessary copies due to having the link loop nested, if we move the link loop as the outer loop then we can copy only once. 

Finally, another minor change, returning a list is reasonable, but let's keep a `dict[str, Link]`, that way if the caller just want to iterate it iterate on `res.values()`, or otherwise it also allow it to perform a `.get(id)` lookup if they want to check if a particular link has been found.

Reply:
The dict being returned looks good. Still, it's making unnecessary extra `self.links` copies.